### ISSUE TITLE
Intel GPU support

### ISSUE THREAD

**Issue by xyang2013** [rasbt/reasoning-from-scratch#19]
https://docs.pytorch.org/docs/stable/notes/get_start_xpu.html

The code in chapter 2 can be modified to:
```
import torch
 
print(f"PyTorch version {torch.__version__}")
if torch.cuda.is_available():
    print("CUDA GPU")
elif torch.xpu.is_available():
    print("Intel GPU")
elif torch.mps.is_available():
    print("Apple Silicon GPU")
else:
    print("Only CPU")
```

**Comment by rasbt** [comment]
Hi there,
yes, it probably could. But I don't have an Intel GPU and don't know how stable the results are compared to the other ones. Have you tried it, and do you get similar outputs as with "cpu"?

**Comment by xyang2013** [comment]
I have used the discrete Intel Arc B580 GPU for your other book (see the video below). It performed quite well.

https://youtu.be/UUrrJWRAb3k?si=WAb9_b0EIUkCL-oE

It should also work with Intel Core Ultra laptops. I tested a simple ResNet model on a first-generation Core Ultra laptop. When plugged in, the iGPU was approximately twice as fast as the CPU. When running on battery, the iGPU offered a significant speedup over the CPU.

**Comment by rasbt** [comment]
Thanks for the info, that sounds pretty good. The one thing I am concerned about is training convergence issues later; but readers with Intel GPU can let me know then 😅. I think it's worthwhile adding it for now due to the 2x speed-up for now.

**Comment by xyang2013** [comment]
I believe so, as that could be significant for people using Intel laptops. As for Intel discrete GPUs, I think they should work, and they are relatively economical to purchase.

If I remember correctly, Intel recently fixed some calculation issues with PyTorch.

**Comment by xyang2013** [comment]
Some training run on the Intel iGPU:

https://x.com/XYang2023/status/1967568894568411310

**Comment by rasbt** [comment]
Thanks for sharing! In my experience, with non-CUDA accelerators, the issue is with small numeric differences that can add up. I.e. this needs to be tested on a case by case basis. For example, some models train fine with "mps" but for some you might get some numeric differences that can add ob so that the model training doesn't converge. E.g., I had observed that in Ch05 (training worked fine) and Ch07 (loss diverged on mps) in my previous book. Other readers reported this too: https://github.com/rasbt/LLMs-from-scratch/discussions/625

So once all the code is done, I hope someone can test it on iGPU so that it can then be officially recommended.

**Comment by xyang2013** [comment]
I should have Intel Arc Pro B60. I can also test it using the discrete GPU vs the iGPU.

**Comment by rasbt** [comment]
Thanks, that'd be cool!
